File: OximyMac/Services/BetterStackLogsService.swift
import Foundation

@MainActor
final class BetterStackLogsService {
    static let shared = BetterStackLogsService()

    private(set) var isInitialized = false
    private var sourceToken: String?
    private var host: String?
    private var session: URLSession?
    private var buffer: [[String: Any]] = []
    private var flushTimer: Timer?

    private let flushInterval: TimeInterval = 5.0
    private let maxBufferSize = 20

    private init() {}

    func initialize() {
        guard let token = Secrets.betterStackLogsToken, !token.isEmpty,
              let host = Secrets.betterStackLogsHost, !host.isEmpty else {
            print("[BetterStackLogs] No token or host configured - Better Stack Logs disabled")
            return
        }

        self.sourceToken = token
        self.host = host

        // Create a URLSession that bypasses mitmproxy
        let config = URLSessionConfiguration.default
        config.connectionProxyDictionary = [:]
        self.session = URLSession(configuration: config)

        isInitialized = true

        flushTimer = Timer.scheduledTimer(withTimeInterval: flushInterval, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                self?.flush()
            }
        }

        print("[BetterStackLogs] Initialized successfully")
    }

    func enqueue(_ entry: [String: Any]) {
        guard isInitialized else { return }

        buffer.append(entry)

        if buffer.count >= maxBufferSize {
            flush()
        }
    }

    func flush() {
        guard isInitialized, !buffer.isEmpty else { return }
        guard let session = session, let host = host, let token = sourceToken else { return }

        let entries = buffer
        buffer = []

        guard let url = URL(string: host),
              let jsonData = try? JSONSerialization.data(withJSONObject: entries, options: []) else {
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData

        session.dataTask(with: request) { _, _, error in
            if let error = error {
                // Fail-open: log locally but don't propagate
                NSLog("[BetterStackLogs] Flush failed: %@", error.localizedDescription)
            }
        }.resume()
    }

    func close() {
        flushTimer?.invalidate()
        flushTimer = nil
        flush()
    }
}

File: OximyWindows/src/OximyWindows/Services/BetterStackLogsService.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using OximyWindows.Core;

namespace OximyWindows.Services;

public static class BetterStackLogsService
{
    private static bool _initialized;
    private static string? _sourceToken;
    private static string? _host;
    private static HttpClient? _httpClient;
    private static readonly ConcurrentQueue<Dictionary<string, object>> _buffer = new();
    private static Timer? _flushTimer;
    private static readonly object _flushLock = new();
    private const int MaxBufferSize = 20;
    private const int FlushIntervalMs = 5000;

    public static bool IsInitialized => _initialized;

    public static void Initialize()
    {
        if (_initialized) return;

        _sourceToken = Secrets.BetterStackLogsToken;
        _host = Secrets.BetterStackLogsHost;

        if (string.IsNullOrEmpty(_sourceToken) || string.IsNullOrEmpty(_host))
        {
            Debug.WriteLine("[BetterStackLogs] No token/host configured - logging disabled");
            return;
        }

        // Bypass proxy to avoid routing through mitmproxy
        var handler = new HttpClientHandler { UseProxy = false };
        _httpClient = new HttpClient(handler);
        _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {_sourceToken}");

        _flushTimer = new Timer(_ => Flush(), null, FlushIntervalMs, FlushIntervalMs);
        _initialized = true;
        Debug.WriteLine("[BetterStackLogs] Initialized successfully");
    }

    public static void Enqueue(Dictionary<string, object> entry)
    {
        if (!_initialized) return;

        _buffer.Enqueue(entry);

        if (_buffer.Count >= MaxBufferSize)
            Flush();
    }

    public static void Flush()
    {
        if (!_initialized || _buffer.IsEmpty) return;

        lock (_flushLock)
        {
            var entries = new List<Dictionary<string, object>>();
            while (_buffer.TryDequeue(out var entry))
                entries.Add(entry);

            if (entries.Count == 0) return;

            _ = SendBatchAsync(entries); // fire and forget
        }
    }

    private static async Task SendBatchAsync(List<Dictionary<string, object>> entries)
    {
        try
        {
            var json = JsonSerializer.Serialize(entries);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            await _httpClient!.PostAsync(_host, content);
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[BetterStackLogs] Send error: {ex.Message}");
            // fail-open: silently swallow
        }
    }

    public static void Close()
    {
        if (!_initialized) return;

        try
        {
            _flushTimer?.Dispose();
            _flushTimer = null;
            Flush(); // send remaining
            _httpClient?.Dispose();
            _httpClient = null;
            _initialized = false;
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[BetterStackLogs] Close error: {ex.Message}");
        }
    }
}
